// @applitools/eyes-webdriverio/dist/api.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchClose = exports.Target = exports.CheckSettings = exports.Configuration = exports.Eyes = exports.By = void 0;
const eyes_sdk_core_1 = require("@applitools/eyes-sdk-core");
const api = __importStar(require("@applitools/eyes-api"));
const spec = __importStar(require("./spec-driver"));
const sdk = eyes_sdk_core_1.makeSDK({
    name: 'eyes.webdriverio',
    version: require('../package.json').version,
    spec,
    VisualGridClient: require('@applitools/visual-grid-client'),
});
__exportStar(require("@applitools/eyes-api"), exports);
var legacy_1 = require("./legacy");
Object.defineProperty(exports, "By", { enumerable: true, get: function () { return legacy_1.By; } });
class Eyes extends api.Eyes {
}
exports.Eyes = Eyes;
Eyes._spec = sdk;
class Configuration extends api.Configuration {
}
exports.Configuration = Configuration;
Configuration._spec = sdk;
class CheckSettings extends api.CheckSettings {
}
exports.CheckSettings = CheckSettings;
CheckSettings._spec = sdk;
exports.Target = CheckSettings;
class BatchClose extends api.BatchClose {
}
exports.BatchClose = BatchClose;
BatchClose._spec = sdk;
